### ISSUE TITLE
Investigate why machinepool annotation typo fix breaks everything

### ISSUE THREAD

**Issue by stbenjam** [metalkube/kni-installer#125]
https://github.com/openshift-metal3/kni-installer/pull/116 caused installs to fail.  We fixed a typo in the JSON annotation, so `openstack` becomes `baremetal`: https://github.com/openshift-metal3/kni-installer/blob/b932498707ed8afd6571db5c419f736be3dbc656/pkg/types/machinepools.go#L61

That means that the output of the code below is now actually used:

https://github.com/openshift-metal3/kni-installer/blob/b932498707ed8afd6571db5c419f736be3dbc656/pkg/asset/machines/baremetal/machines.go#L19

It more or less looks like the other platforms, but this is causing something to happen where the generated assets from dev-scripts which contain all the coredns stuff aren't being used.  Still trying to trace the code to figure out how and why this is happening :confused:

**Comment by stbenjam** [comment]
If you look on the bootstrap node in /opt/openshift/openshift, there's a significant difference. Our generated manifests don't get used anymore (the ones from dev-scripts in assets/generated).


Working (with typo, e.g. baremetal machine pool code not used):

```
opt/openshift/openshift/
├── 99_ingress-controller.yaml
├── 99_kubeadmin-password-secret.yaml
├── 99_master-api-haproxy.yaml
├── 99_master-chronyd-custom.yaml
├── 99_master-coredns.yaml
├── 99_master-core-password.yaml
├── 99_master-dhclient-dns.yaml
├── 99_master-keepalived.yaml
├── 99_master-mdns-publisher.yaml
├── 99_master-utils.yaml
├── 99_openshift-cluster-api_master-machines-0.yaml
├── 99_openshift-cluster-api_master-machines-1.yaml
├── 99_openshift-cluster-api_master-machines-2.yaml
├── 99_openshift-cluster-api_master-user-data-secret.yaml
├── 99_openshift-cluster-api_worker-machineset-0.yaml
├── 99_openshift-cluster-api_worker-user-data-secret.yaml
├── 99_openshift-kni-infra-namespace.yaml
├── 99_openshift-machineconfig_master.yaml
├── 99_openshift-machineconfig_worker.yaml
├── 99_registry.yaml
├── 99_worker-chronyd-custom.yaml
├── 99_worker-coredns.yaml
├── 99_worker-dhclient-dns.yaml
├── 99_worker-mdns-publisher.yaml
└── 99_worker-utils.yaml
```

Non-working:

```
opt/openshift/openshift/
├── 99_kubeadmin-password-secret.yaml
├── 99_openshift-cluster-api_master-machines-0.yaml
├── 99_openshift-cluster-api_master-machines-1.yaml
├── 99_openshift-cluster-api_master-machines-2.yaml
├── 99_openshift-cluster-api_master-user-data-secret.yaml
├── 99_openshift-cluster-api_worker-machineset-0.yaml
├── 99_openshift-cluster-api_worker-user-data-secret.yaml
├── 99_openshift-machineconfig_master.yaml
└── 99_openshift-machineconfig_worker.yaml
```

**Comment by stbenjam** [comment]
/opt/openshift/manifests/cluster-config.yaml are slightly different, there's an empty baremetal hash under contorlPlane["platform"]


Failing:

```yaml
apiVersion: v1
data:
  install-config: |
    apiVersion: v1
    baseDomain: test.metalkube.org
    compute:
    - hyperthreading: Enabled
      name: worker
      platform: {}
      replicas: 1
    controlPlane:
      hyperthreading: Enabled
      name: master
      platform:
        baremetal: {}
      replicas: 3
    metadata:
      creationTimestamp: null
      name: ostest
    networking:
      clusterNetwork:
      - cidr: 10.128.0.0/14
        hostPrefix: 23
      machineCIDR: 10.0.0.0/16
      networkType: OpenShiftSDN
      serviceNetwork:
      - 172.30.0.0/16
    platform:
      baremetal:
        api_vip: 192.168.111.5
        external_bridge: baremetal
        hosts:
        - bmc:
            Address: ipmi://192.168.111.1:6230
            password: password
            username: admin
          bootMACAddress: 00:f5:64:1e:25:3b
          hardwareProfile: default
          name: openshift-master-0
          role: master
        - bmc:
            Address: ipmi://192.168.111.1:6231
            password: password
            username: admin
          bootMACAddress: 00:f5:64:1e:25:3f
          hardwareProfile: default
          name: openshift-master-1
          role: master
        - bmc:
            Address: ipmi://192.168.111.1:6232
            password: password
            username: admin
          bootMACAddress: 00:f5:64:1e:25:43
          hardwareProfile: default
          name: openshift-master-2
          role: master
        image:
          checksum: 2b3b1e19e18627d89da400b63430d5bb
          deployKernel: http://172.22.0.1/images/ironic-python-agent.kernel
          deployRamdisk: http://172.22.0.1/images/ironic-python-agent.initramfs
          source: http://172.22.0.1/images/rhcos-ootpa-latest.qcow2
        ironic_uri: http://localhost:6385/v1
        libvirt_uri: qemu:///system
        provisioning_bridge: provisioning
```

Working

```yaml
apiVersion: v1
data:
  install-config: |
    apiVersion: v1
    baseDomain: test.metalkube.org
    compute:
    - hyperthreading: Enabled
      name: worker
      platform: {}
      replicas: 1
    controlPlane:
      hyperthreading: Enabled
      name: master
      platform: {}
      replicas: 3
    metadata:
      creationTimestamp: null
      name: ostest
    networking:
      clusterNetwork:
      - cidr: 10.128.0.0/14
        hostPrefix: 23
      machineCIDR: 10.0.0.0/16
      networkType: OpenShiftSDN
      serviceNetwork:
      - 172.30.0.0/16
    platform:
      baremetal:
        api_vip: 192.168.111.5
        external_bridge: baremetal
        hosts:
        - bmc:
            Address: ipmi://192.168.111.1:6230
            password: password
            username: admin
          bootMACAddress: 00:ba:7c:33:d7:9a
          hardwareProfile: default
          name: openshift-master-0
          role: master
        - bmc:
            Address: ipmi://192.168.111.1:6231
            password: password
            username: admin
          bootMACAddress: 00:ba:7c:33:d7:9e
          hardwareProfile: default
          name: openshift-master-1
          role: master
        - bmc:
            Address: ipmi://192.168.111.1:6232
            password: password
            username: admin
          bootMACAddress: 00:ba:7c:33:d7:a2
          hardwareProfile: default
          name: openshift-master-2
          role: master
        image:
          checksum: 2b3b1e19e18627d89da400b63430d5bb
          deployKernel: http://172.22.0.1/images/ironic-python-agent.kernel
          deployRamdisk: http://172.22.0.1/images/ironic-python-agent.initramfs
          source: http://172.22.0.1/images/rhcos-ootpa-latest.qcow2
        ironic_uri: http://localhost:6385/v1
        libvirt_uri: qemu:///system
```

**Comment by hardys** [comment]
Ok I looked into this a bit today and gleaned some clues.  The way dev-scripts drives the installer we do:

1. Run kni-installer generate manifests
2. Inject some manifests (the ones missing in the case above)
3. Run kni-installer create cluster (which is supposed to consume the combined manifests from 1 and 2

To debug I replaced the create cluster with create ignition-configs and an exit in dev-scripts, e.g:

```
$ git diff
diff --git a/utils.sh b/utils.sh
index 27478fe..0a8e23e 100644
--- a/utils.sh
+++ b/utils.sh
@@ -50,7 +50,9 @@ function create_cluster() {
     cp -rf assets/generated/*.yaml ${assets_dir}/openshift
 
     cp ${assets_dir}/install-config.yaml{.tmp,}
-    $OPENSHIFT_INSTALLER --dir "${assets_dir}" --log-level=debug create cluster
+    #$OPENSHIFT_INSTALLER --dir "${assets_dir}" --log-level=debug create cluster
+    $OPENSHIFT_INSTALLER --dir "${assets_dir}" --log-level=debug create ignition-configs
+    exit 1
 }
```

Then you can run ```./ocp_cleanup.sh && ./06_create_cluster.sh``` and look at the generated ignition files e.g ocp/bootstrap.ign - we can see the files added in (2) above are missing when we have #116 applied.

The additional manifests are loaded via this code:

https://github.com/openshift-metal3/kni-installer/blob/master/pkg/asset/manifests/openshift.go#L193

However, we see that some dependencies get marked as dirty with #116 applied, so the manifests are re-generated, ignoring those injected in (2) above:

```
# With #116 applied
$ grep -A1000 "create ignition-configs" logs/06_create_cluster-2019-06-21-151445.log | grep "Openshift Manifests"
level=debug msg="  Loading \"Openshift Manifests\"..."
level=warning msg="  Discarding the \"Openshift Manifests\" that was provided in the target directory because its dependencies are dirty and it needs to be regenerated"
level=debug msg="  Fetching \"Openshift Manifests\"..."
level=debug msg="  Generating \"Openshift Manifests\"..."
level=info msg="Consuming \"Openshift Manifests\" from target directory"
level=debug msg="Purging asset \"Openshift Manifests\" from disk"

# Without #116 applied
$ grep -A1000 "create ignition-configs" logs/06_create_cluster-2019-06-21-161354.log | grep "Openshift Manifests"
level=debug msg="  Loading \"Openshift Manifests\"..."
level=debug msg="  Loading \"Openshift Manifests\" from both state file and target directory"
level=debug msg="  Using \"Openshift Manifests\" loaded from target directory"
level=debug msg="  Fetching \"Openshift Manifests\"..."
level=debug msg="  Reusing previously-fetched \"Openshift Manifests\""
level=info msg="Consuming \"Openshift Manifests\" from target directory"
level=debug msg="Purging asset \"Openshift Manifests\" from disk"
```

We can see from the "dependencies are dirty and it needs to be regenerated" message that the loaded manifests are ignored, but I've not yet figured out why the dependencies are considered dirty due to this change.

**Comment by stbenjam** [comment]
Fixed in https://github.com/openshift-metal3/kni-installer/pull/128 and https://github.com/openshift-metal3/dev-scripts/pull/640

**Comment by hardys** [comment]
This has been resolved via #128 and https://github.com/openshift-metal3/dev-scripts/pull/640
